### ISSUE TITLE
Spawn many chains from returned list

### ISSUE THREAD

**Issue by lavalamp-** [celery/celery#3401]
Greetings,

I am attempting to implement the following logic in celery:

Tasks:
- Task A takes in an ID, does a bunch of things, and returns a list of IDs
- Task B takes in an ID, runs a bunch of things

Logic:
- Run Task A, pipe the results to SOMETHING which creates a single Task B for every ID in the list returned by Task A
- Both Task A and Task B's signatures require the population of some key-word arguments, as well as manually setting the queue

From the Celery documentation, I was hoping that I could implement the following:

`Task A | map | Task B`

The implementation of which would look like:

``` py
chain(
    task_a.s().set(queue="foobar"),
    task_b.map(order_id="123", user_id="1234").set(queue="foobar"),
)
```

Clearly, this syntax is NOT supported by the current Celery library. Another approach that I could take is to take the entire contents of the list returned by Task A and feed that list as input to another task which in turn calls xmap, such as:

``` py
@app.task
def my_task(returned_ids, order_id=None, user_id=None):
    return xmap(task_b.s(order_id=order_id, user_id=user_id), returned_ids)
```

However, in the event that Task A returns thousands of IDs (GUIDs, actually) then I'm guessing another component of Celery would break.

Any advice on what way would be best to implement this functionality?

**Comment by ask** [comment]
Are you saying that you want

```py
TaskA.s() | TaskB.map.s()
```

to call map on the results returned by the tasks started by TaskA?

If so this requires synchronization since the return values for the group of tasks TaskA calls
needs to be collected, so it's in effect a chord.

It definitely would be neat if ``B.s()`` would attach itself to the child task if `A.s()` returns a promise (AsyncResult) when `A.s() | B.s()`, and that is even possible, but adding synchronization to that complicates the problem quite a bit.

The primitives in the canvas are all stateless, except for chord which requires a result backend, so there's no global state to update for chains, groups etc.

The `xmap` you're using in your example is unnecessary as you could just as well do:

```py
@app.task
def my_task(returned_ids, order_id=None, user_id=None):
   return map(task_b.s(order_id=order_id, user_id=user_id), returned_ids)
```

This is since all xmap does in your example is to send a new message doing the map operation:

* `group(list_of_n_tasks)` == n task messages
* `task.map(list_of_n_items)` == 1 task message, calling the task as a function on all items.


Now as to your logic, you could solve what you describe in many ways, but the best way is hard to figure without a more concrete example.

You could create a task that takes a callback signature:

```py
@app.task(bind=True)
def TaskA(self, data, callback):
    return group(TaskB.s(item) | self.app.signature(callback)).delay()
```

Calling this with `TaskA.s(data, TaskC.s())` will call TaskB() on every item in data, then call the TaskC callback for every result of TaskB.

You probably recognize that we could generalize this so that TaskB() is also configurable here, that way you have a new primitive for expressing this.

If you want to add synchronization to this you would need to use chords, and then your best option
is to use Redis as a result backend (RabbitMQ as broker, redis for results is a popular combination).
When using Redis as the result backend chords are actually mutable so you can add new tasks to a chord as it executes and other things not possible in the stateless result backends.

Chords work by passing two additional fields to each task message:

1) the `group_id` tells the worker what chord the task belongs to
2) the `chord` header is the serialized callback signature.
   This chord header also has a chord_size attribute which is the number of
   tasks in the chord

When using the Redis result backend you can create specialized synchronization patterns
using these fields, all you need is to pass the headers to every task that should count down
to the callback being called, for example:

```py

from celery import group, uuid


def my_chord(data):
    callback = TaskC.s()
    callback_result = callback.freeze()   # <-- finalizes task_id for this signature, returns AsyncResult
    callback.chord_size = len(data) 
       
    group_id = uuid()

     header = group(
         TaskA.s(item).set(chord=callback, group_id=group_id)
         for item in data
     )
     header.delay()
     return callback_result


Using this information you could create chords that are created at runtime as the workflow executes.

Having said all that, maybe your use case is simply solved by doing:

```py

@app.task(bind=True)
def map_results(self, chord_return_values, callback):
    callback = self.app.signature(callback)  # convert json dict back into Signature
    return [callback(retval) for retval in chord_return_values]

# Collect all return values for TaskA tasks, then call map(TaskC on the return values
chord(TaskA.s(item, order_id=order_id) for item in item)(map_results.s(TaskC.s(order_id=order_id))```   
```

**Comment by lavalamp-** [comment]
Wow - thank you very much for the detailed response. I'm going to write some test code to check that the suggested approach fulfills all the necessary pre-requisites of my issue, but at a glance I bet this solves the problem!

One other approach that I was considering was having a task that does not "finish" until ALL of the child tasks spawned by it have completed. This way a parent task can spawn a bunch of child tasks (which in turn can create their own child tasks, so on and so forth) and "finish" once the entire hierarchy of child tasks have completed. I have a few unknowns right now, which if you have immediate answers for would be greatly appreciated:

(1) If a task in a chord header is retried, does the chord continue to wait until that retried task either successfully completes or errors out?
(2) Can I change a task's type before raising the retry error?

For some additional context on (2), I'd like to have a task that first spawns a bunch of children tasks, and then every 30 seconds checks to see if all those children have successfully completed. One approach would be to have a loop that sleeps for 30 seconds and checks in the parent task's logic, popping out of the loop once the check returns successfully. This approach, however, runs into the same potential issue as calling AsyncResult.get() in a task (consumes a worker process and does nothing with it, potentially tying up all available resources and hitting a deadlock). As such, I'd like to have the initial task "retry," but the logic that is executed upon the retry is only to check that all child result sets have returned, retrying again if the results **have not** returned and marking the task as "completed" if all tasks **have** returned.

If this sounds like something that is not all that crazy, I'd happily add this functionality into the base Task class and submit a pull request. The would result in something that looks like:

`
@app.task(bind=True)
def spawn_the_children(self, how_many):
    results = group([other_task.s() for i in range(how_many)])  # Spawns child processes
    self.complete_if_results_finished(results)  # Simple call that does a bunch of things in the back-end, but ensures this task is not "finished" until children and children's children are finished.
`

**Comment by linar-jether** [comment]
If this is still relevant, this is what we use to do what you want, it is inspired by (http://stackoverflow.com/questions/13271056/how-to-chain-a-celery-task-that-returns-a-list-into-a-group).

We use it like this: 
*Note this example executes a pipeline of dynamically created tasks (serialized using dill)
```python
# Task primitives, allows pipeline execution using celery
@app.task
def dmap(it, callback, final=None):
    # Map a callback over an iterator and return as a group
    callback = subtask(callback)

    # Hack for mapping a chain to values, due to a bug where args are not copied in group creation
    if isinstance(callback, chain):
        if final:
            raise ValueError('task_processor: Cannot run reducer for dmap excecuted with a chain.')

        return [callback.delay(arg) for arg in it]

    run_in_parallel = group(callback.clone([arg, ]) for arg in it)

    if len(run_in_parallel.tasks) == 0:
        return []

    if final:
        return chord(run_in_parallel)(final)

    return run_in_parallel.delay()

@app.task
def _exec_async(*args, **kwargs):
    import dill
    import zlib

    serialized_callable = kwargs.pop('_serialized_callable', None)

    if serialized_callable is None:
        raise Exception('Missing serialized_callable kwarg, must contain the dill-serialized callable.')

    # Maybe compressed
    try:
        serialized_callable = serialized_callable.decode('zlib')
    except zlib.error:
        pass

    return dill.loads(serialized_callable)(*args, **kwargs)


def exec_async(func, queue=None, *args, **kwargs):
    import dill
    queue = queue

    if queue is None:
        raise ValueError("Missing queue name")

    serialized_callable = dill.dumps(func)
    kwargs['_serialized_callable'] = serialized_callable
    args_ = list(args)

    if kwargs.pop('_as_sig', False):
        return _exec_async.s(*args_, **kwargs).set(queue=queue)

    return _exec_async.apply_async(args_, kwargs, queue=queue, compression='zlib')


def async(func, queue=None, *args, **kwargs):
    kwargs['_as_sig'] = True

    return exec_async(func, queue, *args, **kwargs)


# DAG components

def get_list_of_items():
    return range(100)

def do_something_for_item(item):
    import numpy as np
    return item * np.random.rand()

def get_top_10(items):
    import pandas as pd
    return pd.Series(items).sort_values()[-10:]

queue_name = 'test_queue'

dag = async(get_list_of_items, queue_name) | async(dmap, queue_name, callback=async(do_something_for_item, queue_name), final=async(get_top_10, queue_name))

# Execute DAG
dag.delay().get().get()
```

**Comment by Sovetnikov** [comment]
@lavalamp- I just read this great thread, but how you make "Task A | map | Task B" work with celery?
And how you solved problem with thousands of ids returned by TaskA? Have you implemented dynamic tasks insertion to chord header?

**Comment by lavalamp-** [comment]
Hey @Sovetnikov - yeah I actually did find a solution for this problem. Basically I have a list of "tags" that are included in task request headers. These tags are incremented and decremented in a Redis instance as tasks are queued up / finished / errored, and the parent task that kicks them off morphs itself into a separate task type that retries every 10 seconds, checks to see if the tag in Redis is == 0, and finishes if it is.

Sorry for the poor explanation - here is the code that you can start tracing if you'd like to implement something similar:

https://github.com/lavalamp-/ws-backend-community/blob/master/tasknode/tasks/base.py#L110-L130

Basically the flow using the function highlighted above is to call `self.finish_after(signature=<signature>)` and the task that calls that function won't be "finished" until the tasks defined in the given signature have completed. Furthermore due to the morphing into a different task type and using retries we are not tying up a process checking to see if the children have finished.

**Comment by sposs** [comment]
I'd like to go back to the explanation provided by @ask. Indeed, I'm very must facing the problem myself with celery v4.1.0. I have a task that produces many images that I need to upload in parallel, and after the upload is completed I need to cleanup. Essentially, my chain is 
```python
my_chain = chord(download_group, merge_async.s(out_path)) | \
            cleanup_async.si(out_path) | \
            handle_images.si(args) | \
            upload_and_register | finalize_update_task.s() | \
            final_cleanup_async.si(out_path)
```
The issue is that I do not understand how to write my `upload_and_register` task that handles the output of `handle_images` which is a `list` of local files to be uploaded. I tried something like

```python
@app.task(bind=True)
def dmap(self, it, callback):
    # Map a callback over an iterator and return as a group
    callback = self.app.signature(callback)
    return group(callback(arg) for arg in it)()
```
and my chain looks like
```python
my_chain = chord(download_group, merge_async.s(out_path)) | \
            cleanup_async.si(out_path) | \
            handle_images.si(args) | \
            dmap.s(upload_and_register.s()) | finalize_update_task.s() | \
            final_cleanup_async.si(out_path)
```
where `upload_and_register` handles only one file. The problem is that the tasks defined in `dmap` are executed serially, while I want to have my upload_and_register run in parallel (as much as possible). I think the 
```python
def my_chord(data):
```
method looks like what I need, but I have no idea how to integrate it... Is it a task? How is is called in the chain?

I understand that some synchronization is needed before the chain can proceed (it's a chord that I need somewhere, but I guess the chain for a group with something makes a chord...). And that when the chain is defined, it's not obvious to tell the system to wait and see, especially given that the tasks are not all defined when the chain is `delay`ed. Any idea would be most welcome, I've been banging my head and thinking I should have a separate system to handle the file uploads...

Thanks

**Comment by Sovetnikov** [comment]
@sposs, IMO dmap (and analogous task logic implementation) is not production ready in celery ...
"Advanced" task chaining (beyond task A | task B | task C | task ...) is not what celery is aimed to.
I have similar demand to chain tasks with lots of forks and joins, and i'm now looking at dask and dask distributed to make it work (https://distributed.readthedocs.io/en/latest/)

**Comment by linar-jether** [comment]
@sposs, See my comment above, we're achieving this by pushing the task signatures (i.e. the join part) as arguments to the `dmap` function.

```python
@app.task
def dmap(it, callback, final=None):
    # Map a callback over an iterator and return as a group
    callback = subtask(callback)

    # Hack for mapping a chain to values, due to a bug where args are not copied in group creation
    if isinstance(callback, chain):
        if final:
            raise ValueError('task_processor: Cannot run reducer for dmap excecuted with a chain.')

        return [callback.delay(arg) for arg in it]

    run_in_parallel = group(callback.clone([arg, ]) for arg in it)

    if len(run_in_parallel.tasks) == 0:
        return []

    if final:
        return chord(run_in_parallel)(final)

    return run_in_parallel.delay()

# DAG components

@app.task
def get_list_of_items():
    return range(100)

@app.task
def do_something_for_item(item):
    import numpy as np
    return item * np.random.rand()

@app.task
def get_top_10(items):
    import pandas as pd
    return pd.Series(items).sort_values()[-10:]


dag = get_list_of_items.s() | dmap.s(callback=do_something_for_item.s(), final=get_top_10.s())

# Execute DAG
dag.delay()
```

**Comment by sposs** [comment]
Small correction:
```python
return chord(run_in_parallel)(final)
```
should be 
```python
return chord(run_in_parallel)(subtask(final))
```

But otherwise it works absolutely perfectly! I've been banging my head for a clear answer and this one is excellent. Thanks!

**Comment by sposs** [comment]
Final comment for other users: the `final` must contain the rest of the chain, in my case:
```python
 dmap.s(upload_and_register.s()) final=(finalize_update_task.s() | final_cleanup_async.si(out_path)))
```
Fortunately, celery allows defining chains a priori, so it's easy to create one, even complex and pass it in. This stuff is really great.

**Comment by quantoid** [comment]
But in this solution, dmap is a task that invokes a group of tasks, which I thought was an antipattern for Celery tasks. 

In other solutions I’ve seen, dmap returns the group without executing it. Does that mean a task can return a signature, and that signature will be called by Celery to get the true return value of that task?

**Comment by jrectorb** [comment]
I'd also be interested in an answer to @quantoid's question.  

I'm in a similar situation to those already brought up in this thread, and the natural solution seems to calling a group of tasks within another subtask as in the solution with `dmap`.  Everything I've read, though, labels this sort of a solution an antipattern.  Is there something I'm missing?  Is there something going on behind the scenes which makes such a sequence of calls okay?  If not, is it true (like @quantoid asked) that a task can itself return a collection of signatures which will then be called by celery to get the return values of those tasks?

**Comment by virtualxdriver** [comment]
As far as I understand, synchronously waiting for the results of a sub task using `get()` is an anti pattern in celery. Starting a sub task by itself is OK.
With that said, the main issue with the above dmap implementation, is that celery will not consider the started sub tasks as part of the initial chain, so to fetch the results, you have to rely to awkward gymnastics such as calling `get()` twice followed by a `join()` on the returned AsyncResult object.

**Comment by huogerac** [comment]
> @sposs, See my comment above, we're achieving this by pushing the task signatures (i.e. the join part) as arguments to the `dmap` function.
> 
> ```python
> @app.task
> def dmap(it, callback, final=None):
>     # Map a callback over an iterator and return as a group
>     callback = subtask(callback)
> 
>     run_in_parallel = group(callback.clone([arg, ]) for arg in it)
> 
>     if len(run_in_parallel.tasks) == 0:
>         return []
> 
>     if final:
>         return chord(run_in_parallel)(final)
> 
>     return run_in_parallel.delay()
> 
> # DAG components
> 
> @app.task
> def get_list_of_items():
>     return range(100)
> 
> @app.task
> def do_something_for_item(item):
>     import numpy as np
>     return item * np.random.rand()
> 
> @app.task
> def get_top_10(items):
>     import pandas as pd
>     return pd.Series(items).sort_values()[-10:]
> 
> 
> dag = get_list_of_items.s() | dmap.s(callback=do_something_for_item.s(), final=get_top_10.s())
> 
> # Execute DAG
> dag.delay()
> ```

Hello,

I believe I had the same Issue where I'd like to create new messages (tasks for  workers) based on my first task that returns a list of items:
TaskA_get_list_of_items
TaskB_do_something_for_item

The code using the dmap works, however, it seems a  lot of things, indeed we need to use .get().get()...
I've tried another approach and it seems to be working:

What you guys think this solution?
Cheers

`
# my service.py
signature = (
    tasks.get_list_of_items.s(input=42) |
    tasks.dynamic_map.s(
        callback=tasks.do_something_for_item,
    )
).delay()

result = signature.get()
`

and the celery tasks:
`
# tasks.py

@app.task
def dynamic_map(it, callback):
    """ creates a new message queue for each item from it """    
    signature = (
        callback.map(it)
    )()
    return signature

@app.task
def get_list_of_items(input):
    ...
    return [...] # list of items

@app.task
def do_something_for_item(item):
    # process one item

`

**Comment by Kxrr** [comment]
> @sposs, See my comment above, we're achieving this by pushing the task signatures (i.e. the join part) as arguments to the `dmap` function.
> 
> ```python
> @app.task
> def dmap(it, callback, final=None):
>     # Map a callback over an iterator and return as a group
>     callback = subtask(callback)
> 
>     run_in_parallel = group(callback.clone([arg, ]) for arg in it)
> 
>     if len(run_in_parallel.tasks) == 0:
>         return []
> 
>     if final:
>         return chord(run_in_parallel)(final)
> 
>     return run_in_parallel.delay()
> 
> # DAG components
> 
> @app.task
> def get_list_of_items():
>     return range(100)
> 
> @app.task
> def do_something_for_item(item):
>     import numpy as np
>     return item * np.random.rand()
> 
> @app.task
> def get_top_10(items):
>     import pandas as pd
>     return pd.Series(items).sort_values()[-10:]
> 
> 
> dag = get_list_of_items.s() | dmap.s(callback=do_something_for_item.s(), final=get_top_10.s())
> 
> # Execute DAG
> dag.delay()
> ```

I think the problem with the **dmap** function above is that it send the map task to celery, but you can't **track** them as you expect.

You can't get the map result by calling `dag.delay().get()`, you will get a AsyncResult object instead.

And when ``dag.delay().state == "SUCCESS"``, the child map tasks could be still running, you will need

```
result = dag.delay()
sub_tasks_result = result.get()
result.state == "SUCCESS" and sub_tasks_result.state == "SUCCESS"
```
to ensure all the tasks is done.
